analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterValidationInMethodWalker.cs

```diff
@@ -44,36 +44,39 @@ internal class ParameterValidationInMethodWalker : SafeCSharpSyntaxWalker
 
         public IEnumerable<Location> ArgumentExceptionLocations => argumentExceptionLocations;
 
-        public ParameterValidationInMethodWalker(SemanticModel semanticModel)
-        {
-            this.semanticModel = semanticModel;
-        }
+        public ParameterValidationInMethodWalker(SemanticModel semanticModel) => this.semanticModel = semanticModel;
 
         public override void Visit(SyntaxNode node)
         {
-            if (this.keepWalking &&
-                !node.IsAnyKind(SubMethodEquivalents))  // Don't explore deeper if this node is equivalent to a method declaration
+            if (keepWalking
+                && !node.IsAnyKind(SubMethodEquivalents))  // Don't explore deeper if this node is equivalent to a method declaration
             {
                 base.Visit(node);
             }
         }
 
-        public override void VisitAwaitExpression(AwaitExpressionSyntax node)
-        {
-            this.keepWalking = false;
-        }
+        public override void VisitAwaitExpression(AwaitExpressionSyntax node) => keepWalking = false;
 
         public override void VisitThrowStatement(ThrowStatementSyntax node)
         {
             // When throw is like `throw new XXX` where XXX derives from ArgumentException, save location
-            if (node.Expression is ObjectCreationExpressionSyntax oces &&
-                this.semanticModel.GetSymbolInfo(oces.Type).Symbol?.OriginalDefinition is ITypeSymbol typeSymbol &&
-                typeSymbol.DerivesFrom(KnownType.System_ArgumentException))
+            if (node.Expression is ObjectCreationExpressionSyntax oces
+                && semanticModel.GetSymbolInfo(oces.Type).Symbol?.OriginalDefinition is ITypeSymbol typeSymbol
+                && typeSymbol.DerivesFrom(KnownType.System_ArgumentException))
             {
-                this.argumentExceptionLocations.Add(oces.GetLocation());
+                argumentExceptionLocations.Add(oces.GetLocation());
             }
 
             base.VisitThrowStatement(node);
         }
+
+        public override void VisitInvocationExpression(InvocationExpressionSyntax node)
+        {
+            if (node.IsMemberAccessOnKnownType("ThrowIfNull", KnownType.System_ArgumentNullException, semanticModel))
+            {
+                argumentExceptionLocations.Add(node.GetLocation());
+            }
+            base.VisitInvocationExpression(node);
```

Comment:
There is no need to visit the arguments.
```suggestion
```

Reply:
`ArgumentNullException.ThrowIfNull` is a throw helper and the first argument is annotated with `[NotNull]`
```cs
public static void ThrowIfNull([NotNull] object? argument, [CallerArgumentExpression("argument")] string? paramName = null);
```
We should detect such throw helpers by looking for `[NotNull]` on arguments, and `[DoesNotReturnIf]` and `[DoesNotReturn]` on invocations. This is the more robust way to handle this.

[Reference](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/attributes/nullable-analysis) 

Add some tests with custom throw helpers.

Reply:
I'd say having more complex logic should require a different ticket.

Just checking `ThrowIfNull` is ok from a performance point of view because it first checks the name and only then does a semantic check. 

Having a catch-all logic for throw helpers might be more expensive and needs it's own thought.

I've opened https://github.com/SonarSource/sonar-dotnet/issues/6368 for it.

This PR is only about `ArgumentNullException.ThrowIfNull()` which comes out of the box and is more widely used, I won't add test cases here for the more generic case.